mautic/mautic
# Command to deduplicate contacts

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? | y
| Related user documentation PR URL | todo
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds a command `php app/console mautic:contacts:dedup` to find contacts with the same unique identifiers and merge them. By default, the records are merged from oldest into newest but that can be changed with the flag `--newest-into-oldest`. 

Regardless of which record is merged into which, the latest modified record's custom field data is used (except where it is empty as we will error on the side of not completely losing data). Also the latest last active date is preserved and the oldest date identified date. 

Contacts with conflicting unique identifiers when there are more than one unique identifier are not merged. 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Hack the database to update multiple contacts with the same email OR manually edit through the UI to ensure there are multiple contacts with the same unique identifiers. 
2. Run the command and they should be merged along with all of their behavior data. 

#### List deprecations along with the new alternative:
1. LeadModel::mergeLeads is deprecated. ContactMerger should be used instead
